sonic_y_cable/y_cable.py

```diff
@@ -1097,11 +1128,135 @@ def check_if_nic_lanes_active(physical_port):
 
     if platform_chassis is not None:
         res = platform_chassis.get_sfp(physical_port).read_eeprom(curr_offset, 1)
-        y_cable_validate_read_data(res, 1, physical_port, "nic lanes active")
+        if y_cable_validate_read_data(res, 1, physical_port, "nic lanes active") == EEPROM_READ_DATA_INVALID:
+            return EEPROM_ERROR
         result = res[0]
 
     else:
         helper_logger.log_error("platform_chassis is not loaded, failed to get NIC lanes active")
         return -1
 
     return result
+
+
+def get_firmware_version(physical_port, target):
+
+    data = bytearray(FIRMWARE_INFO_PAYLOAD_SIZE)
+
+    if platform_chassis is not None:
+        for byteIdx in range(0, FIRMWARE_INFO_PAYLOAD_SIZE):
+            curr_offset = 0xfc * 128 + 128 + byteIdx
+            readOut = platform_chassis.get_sfp(physical_port).read_eeprom(curr_offset, 1)
+            if y_cable_validate_read_data(readOut, 1, physical_port, "firmware info") == EEPROM_READ_DATA_INVALID:
+                return EEPROM_ERROR
+            data[byteIdx] = readOut[0]
+    else:
+        helper_logger.log_error("platform_chassis is not loaded, failed to get NIC lanes active")
+        return -1
+
+    result = {}
+    NUM_MCU_SIDE = 3
+    base_addr = int(target * (FIRMWARE_INFO_PAYLOAD_SIZE / NUM_MCU_SIDE))
+    rev_major_slot1 = struct.unpack_from('<B', data[0 + base_addr:  1 + base_addr])[0]
+    rev_minor_slot1 = struct.unpack_from('<B', data[2 + base_addr:  3 + base_addr])[0]
+    rev_build_lsb_slot1 = struct.unpack_from('<B', data[4 + base_addr:  5 + base_addr])[0]
+    rev_build_msb_slot1 = struct.unpack_from('<B', data[5 + base_addr:  6 + base_addr])[0]
+    rev_major_slot2 = struct.unpack_from('<B', data[1 + base_addr:  2 + base_addr])[0]
+    rev_minor_slot2 = struct.unpack_from('<B', data[3 + base_addr:  4 + base_addr])[0]
+    rev_build_lsb_slot2 = struct.unpack_from('<B', data[6 + base_addr:  7 + base_addr])[0]
+    rev_build_msb_slot2 = struct.unpack_from('<B', data[7 + base_addr:  8 + base_addr])[0]
+    slot_status = struct.unpack_from('<B', data[8 + base_addr:  9 + base_addr])[0]
```

Comment:
Regarding style, there should be no spaces after `:` for list slicing. Suggest the following:

```python
rev_major_slot1 = struct.unpack_from('<B', data[(0 + base_addr):(1 + base_addr)])[0]
rev_minor_slot1 = struct.unpack_from('<B', data[(2 + base_addr):(3 + base_addr)])[0]
...
```

Reply:
fixed

Reply:
Thanks. I see you didn't add parentheses around the slice values. While not necessary, I added them in my suggestion because I felt they made it more readable. What do you think?

Reply:
fixed, its more readable now